fusesoc/capi2/exprs.py

```diff
@@ -0,0 +1,203 @@
+    (negated, flag, exprs).
+
+    Here, negated is a boolean which is true if the condition should be
+    negated. flag is a word giving the flag to match for the condition. exprs
+    is the AST for the list of exprs inside the parentheses.
+
+    To avoid creating the parser repeatedly, this function is memoized.
+
+    """
+    global _PARSER
+    if _PARSER is not None:
+        return _PARSER
+
+    word = Word(alphanums + ":<>.[]_-,=~/^~")
+    exprs = Forward()
+
+    conditional = (
+        Optional("!")
+        + word
+        + Suppress("?")
+        + Suppress("(")
+        + Group(exprs)
+        + Suppress(")")
+    )
+    exprs <<= OneOrMore(conditional ^ word)
+    conditional.setParseAction(_cond_parse_action)
+    _PARSER = exprs
+    return _PARSER
+
+
+def _simplify_ast(raw_ast):
+    """Simplify an AST that comes out of the parser
+
+    As well as replacing pyparsing's ParseResults with bare lists, this merges
+    adjacent non-condition words . For example, "a b" parses to ["a", "b"].
+    This function merges that to ["a b"].
+
+    The idea is that this will be much more efficient to match against tags for
+    the vast majority of ASTs, which have many more raw words than they have
+    conditions.
+
+    A simplified AST is a list whose items are strings (representing bare
+    words) or tuples of the form (negated, flag, ast), where negated is a bool,
+    flag is a string and ast is another simplified ast.
+
+    """
+    children = []
+    str_acc = []
+    for expr in raw_ast:
+        if isinstance(expr, str):
+            str_acc.append(expr)
+            continue
+
+        # We have a term that isn't a string. This must be a conditional. Join
+        # together any items in str_acc and add them to children then recurse
+        # to simplify the conditional's sub-expression.
+        if str_acc:
+            children.append(" ".join(str_acc))
+            str_acc = []
+
+        negated, flag, exprs = expr
+        children.append((negated, flag, _simplify_ast(exprs)))
+
+    if str_acc:
+        children.append(" ".join(str_acc))
+
+    return children
+
+
+def _parse(string):
+    """Parse a string to a simplified AST.
+
+    Raises a ValueError if the string is malformed in some way.
+
+    """
+    try:
+        raw_ast = _get_parser().parseString(string, parseAll=True)
+    except ParseException as err:
+        raise ValueError(
+            "Invalid syntax for string: {}. Parsed text was {!r}.".format(err, string)
+        ) from None
+
+    return _simplify_ast(raw_ast)
+
+
+class Exprs:
+    """A parsed list of exprs"""
+
+    def __init__(self, string):
+        self.ast = _parse(string)
+        self.as_string = None
+
+        # An extra optimisation for the common case where the whole ast boils
+        # down to a single string with no conditions.
+        if len(self.ast) == 1 and isinstance(self.ast[0], str):
+            self.as_string = self.ast[0]
+
+    @staticmethod
```

Comment:
Maybe, but I prefer explicitly not passing `self` if there isn't one needed. Especially in Python, you're explicitly not passing an unused argument, which I think is better. (It's less obvious in e.g. C++ because the syntax rarely needs `this` explicitly)